src/base/misc/fatfs.c

```diff
@@ -1438,242 +1437,171 @@ unsigned next_cluster(fatfs_t *f, unsigned clu)
-
-  t_o =  0x40 + (boot_message-boot_prog);
-
-  /* update the message */
-  snprintf((char *)b + t_o, eos - t_o, msg_f, f->dir);
 
   /* 1st root directory sector */
   r_o = f->fats * f->fat_secs + f->reserved_secs + f->hidden_secs;
 
   /* 1st data sector (= start of 1st system file) */
   d_o = r_o + f->root_secs;
 
-  txfr->drive = f->drive_num;
+  /* defaults */
+  seg = 0x0070;
+  ofs = 0x0000;
+  loadaddress = SEGOFF2LINEAR(seg, ofs);
+  size = f->obj[1].size;
 
   switch(f->sys_type) {
     case NEWMSD_D:                     /* for IO.SYS, MS-DOS version >= 7 */
-      txfr->ofs = 0x0200;
-      txfr->seg = 0x0070;
-      txfr->ax  = d_o & 0xffff;
-      txfr->dx  = d_o >> 16;
-      txfr->di  = 0x0002;
-      txfr->bp  = 0x7c00;
-
-      txfr->list_entries = 1;
-      make_i1342_blk(&txfr->dlist[0], d_o, 4, 0x70, 0);
-
-      /*
-       * IO.SYS normally re-uses the boot block's error message. We
-       * could give it a distinct one here (simply have t_o point to it).
-       *
-       * But don't forget, it's not a simple ASCIIZ string!!!
-       *
-       * NOTE: There is a discrepancy between IO.SYS's and the bootblock's
-       * interpretation of this address. IO.SYS expects t_o to be
-       * relative to 0, but the value actually stored is relative to 0x7c00.
-       * (Leading IO.SYS to display no error message.)
-       * I will assume IO.SYS to be correct for now.
-       */
-      b[0x1ee] = (0x7c00 + t_o);                  /* address of error msg */
-      b[0x1ef] = (0x7c00 + t_o) >> 8;
-      fatfs_msg("made boot block suitable for MS-DOS, version >= 7\n");
+      seg = 0x0070;
+      ofs = 0x0200;
+      loadaddress = SEGOFF2LINEAR(seg, 0x0000); // different
+      size = 4 * SECTOR_SIZE;
+
+      LWORD(eax) = d_o & 0xffff;
+      LWORD(edx) = d_o >> 16;
+      LWORD(edi) = 0x0002;
+      LWORD(ebp) = 0x7c00;
       break;
 
     case MIDMSD_D:
     case NEWPCD_D:		/* MS-DOS 4.0 -> 6.22 / PC-DOS 4.0 -> 7.0 */
-      txfr->ofs = 0x0000;
-      txfr->seg = 0x0070;
-      txfr->ax  = d_o >> 16;
-      txfr->bx  = d_o & 0xffff;
-      txfr->cx  = f->media_id << 8;   /* ch */
-      txfr->dx  = f->drive_num;
-
-      txfr->list_entries = 2;
-      make_i1342_blk(&txfr->dlist[0], r_o, 1, 0x50, 0);
-      make_i1342_blk(&txfr->dlist[1], d_o, 4, 0x70, 0);
-
-      fatfs_msg("made boot block suitable for MS-DOS 4.0 -> 6.22 & PC-DOS v4.0 -> v7.0\n");
+      // load root directory to 0000:0500
+      read_root(f, 0, LINEAR2UNIX(SEGOFF2LINEAR(0x0000, 0x0500)));
+
+      size = 4 * SECTOR_SIZE;
+
+      LWORD(eax) = d_o >> 16;
+      LWORD(ebx) = d_o & 0xffff;
+      LWORD(ecx) = f->media_id << 8;   /* ch */
+      LWORD(edx) = f->drive_num;
       break;
 
     case OLDPCD_D:		/* old MS-DOS & PC-DOS < v4.0 */
     case OLDMSD_D:
-      txfr->ofs = 0x0000;
-      txfr->seg = 0x0070;
+      // load root directory to 0000:0500
+      read_root(f, 0, LINEAR2UNIX(SEGOFF2LINEAR(0x0000, 0x0500)));
```

Comment:
IIRC in various docs it is spelled as 50:0, not 0:500.
It is usually better to use normalized pointers where
possible.
This is certainly minor and can stay.